packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1243 @@
+            return true;
+        }
+        return false;
+    }
+
+    private focusCurrentRow(allowScroll: boolean): boolean {
+        const visibleRowIndex = this.getVisibleRowIndex();
+        if (visibleRowIndex < 0) {
+            return false;
+        }
+        const focusedRow = this.table.rowElements[visibleRowIndex]!;
+
+        let focusRowOnly = false;
+        switch (this.focusType) {
+            case TableFocusType.row:
+                focusRowOnly = true;
+                break;
+            case TableFocusType.cell:
+            case TableFocusType.cellActionMenu:
+            case TableFocusType.cellContent:
+                focusRowOnly = focusedRow instanceof TableGroupRow;
+                break;
+            default:
+                break;
+        }
+        const focusOptions = { preventScroll: !allowScroll };
+        if (focusRowOnly) {
+            this.focusElement(focusedRow, focusOptions);
+            return true;
+        }
+        this.focusRowElement(focusedRow, focusOptions);
+        return true;
+    }
+
+    private focusRowElement(
+        row: TableRow | TableGroupRow,
+        focusOptions?: FocusOptions
+    ): void {
+        const rowElements = row.getFocusableElements();
+        let focusableElement: HTMLElement | undefined;
+        switch (this.focusType) {
+            case TableFocusType.rowSelectionCheckbox:
+                focusableElement = rowElements.selectionCheckbox;
+                break;
+            case TableFocusType.cell: {
+                focusableElement = rowElements.cells[this.columnIndex]!.cell;
+                break;
+            }
+            case TableFocusType.cellActionMenu:
+                focusableElement = rowElements.cells[this.columnIndex]?.actionMenuButton;
+                break;
+            case TableFocusType.cellContent: {
+                focusableElement = rowElements.cells[this.columnIndex]?.cell.cellView
+                    .tabbableChildren[this.cellContentIndex];
+                break;
+            }
+            default:
+                break;
+        }
+        if (focusableElement) {
+            this.focusElement(focusableElement, focusOptions);
+        }
+    }
+
+    private focusHeaderElement(): boolean {
+        const headerElements = this.getTableHeaderFocusableElements();
+        let focusableElement: HTMLElement | undefined;
+        switch (this.focusType) {
+            case TableFocusType.headerActions:
+                focusableElement = headerElements.headerActions[this.headerActionIndex];
+                break;
+            case TableFocusType.columnHeader:
+                focusableElement = headerElements.columnHeaders[this.columnIndex];
+                break;
+            default:
+                break;
+        }
+        if (focusableElement) {
+            this.focusElement(focusableElement);
+            return true;
+        }
+        return false;
+    }
+
+    private getVisibleRowIndex(): number {
```

Comment:
Should this be `getVisibleFocusedRowIndex()`, or even just `getFocusedRowIndex()`? The current term doesn't provide enough intent in the semantics, as there should be many visible rows at any given time.

Reply:
Renamed to `getCurrentRowVisibleIndex()` for now, which I think makes sense since it's used in `getCurrentRow()`.

I did struggle a bit with some of the method naming. 'Focused row' isn't necessarily accurate, as it may not yet be focused (but is about to be focused, generally).